### PR TITLE
Timed tasks automatically synchronize the extension information of the git address

### PR DESCRIPTION
**Pull request by kakj-go**
#### What type of this PR
/kind feature

#### What this PR does / why we need it:
Timed tasks automatically synchronize the extension information of the git address

#### Which issue(s) this PR fixes:
erda-issue: [erda-issue](https://erda.cloud/erda/dop/projects/387/issues/all?id=209340&issueFilter__urlQuery=eyJpdGVyYXRpb25JRHMiOls1MDZdLCJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIiwiUkVTT0xWRUQiXSwiYXNzaWduZWVJRHMiOlsiMTAwMDU2MCJdfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASK)

### CODE REVIEW

File: modules/dicehub/extension/action.go [hunk 1]
```diff
@@ -49,7 +51,46 @@ type Version struct {
...
+func (s *extensionService) pushGitExtensions(gitAddr string) error {
+	dir, err := ioutil.TempDir(os.TempDir(), "*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	// git init
+	command := exec.Command("sh", "-c", "git clone "+gitAddr)
+	command.Dir = dir
```
Comment (sfwn) [inline modules/dicehub/extension/action.go:63]: handle stdout&stderr correctly.

File: modules/dicehub/extension/action.go [hunk 2]
```diff
@@ -49,7 +51,46 @@ type Version struct {
...
+	err = command.Run()
+	if err != nil {
+		return err
+	}
+
+	return s.InitExtension(dir, true)
+}
+
+func (s *extensionService) TimeTaskSynchronizationExtensions() {
```
Comment (sfwn) [inline modules/dicehub/extension/action.go:72]: timed

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/dicehub/extension/provider.go [hunk 3]
```diff
@@ -32,7 +32,8 @@ import (
 )
 
 type config struct {
-	ExtensionMenu map[string][]string `file:"extension_menu" env:"EXTENSION_MENU"`
+	ExtensionMenu    map[string][]string `file:"extension_menu" env:"EXTENSION_MENU"`
+	ExtensionGitAddr string              `file:"extension_git_addr" env:"EXTENSION_GIT_ADDR"`
```
Comment (Effet) [inline modules/dicehub/extension/provider.go:36]: `extension_sources` is better? What do you think? @sfwn

File: modules/dicehub/extension/action.go [hunk 4]
```diff
@@ -49,7 +51,47 @@ type Version struct {
...
+		if err != nil {
+			logrus.Errorf("error to sync git address %s extension", gitAddr)
+		}
+	}
+
+	logrus.Infof("end to TimeTaskSynchronizationExtensions")
+
+	time.AfterFunc(24*time.Hour, func() {
+		go s.TimedTaskSynchronizationExtensions()
+	})
```
Comment (Effet) [inline modules/dicehub/extension/action.go:91]: why not to set timer out of func?

File: modules/dicehub/extension/action.go [hunk 5]
```diff
@@ -49,7 +51,47 @@ type Version struct {
 	SwaggerContent []byte // content of swagger.yml
 }
 
-func (s *extensionService) InitExtension(addr string) error {
+func (s *extensionService) pushGitExtensions(gitAddr string) error {
+	dir, err := ioutil.TempDir(os.TempDir(), "*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
```
Comment (Effet) [inline modules/dicehub/extension/action.go:59]: It's better to store in suitable place, and do git pull after cloned. I think you'd better add TODO to describe this plan, you can do it later.

**Review by Effet — DISMISSED** [review]

File: modules/dicehub/extension/action.go (hunk 5 above)
Comment (kakj-go) [reply modules/dicehub/extension/action.go:58]: ok

File: modules/dicehub/extension/provider.go [hunk 6]
```diff
@@ -61,12 +63,26 @@ func (p *provider) Init(ctx servicehub.Context) error {
...
+
+	go func() {
+		c := cron.New()
+
+		err := c.AddFunc("0 0 21 * * ?", func() {
+			go p.extensionService.TimedTaskSynchronizationExtensions()
+		})
+
+		if err != nil {
+			logrus.Errorf("error to add cron task")
```
Comment (sfwn) [inline modules/dicehub/extension/provider.go:81]: err was ate.

File: modules/dicehub/extension/provider.go [hunk 7]
```diff
@@ -61,12 +63,26 @@ func (p *provider) Init(ctx servicehub.Context) error {
...
 		if err != nil {
 			panic(err)
 		}
 		logrus.Infoln("End init extension")
 	}()
+
+	go func() {
+		c := cron.New()
+
+		err := c.AddFunc("0 0 21 * * ?", func() {
```
Comment (sfwn) [inline modules/dicehub/extension/provider.go:76]: should be configurable.

File: modules/dicehub/extension/provider.go [hunk 8]
```diff
@@ -61,12 +63,26 @@ func (p *provider) Init(ctx servicehub.Context) error {
...
+			go p.extensionService.TimedTaskSynchronizationExtensions()
+		})
+
+		if err != nil {
+			logrus.Errorf("error to add cron task")
+		}
+
+		c.Start()
```
Comment (sfwn) [inline modules/dicehub/extension/provider.go:84]: invalid start if err occurred.

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**